akka-typed/src/main/scala/akka/typed/cluster/sharding/ClusterShardingSettings.scala

```diff
@@ -0,0 +1,282 @@
+  val coordinatorSingletonSettings: ClusterSingletonManagerSettings) extends NoSerializationVerificationNeeded {
+
+  import akka.typed.cluster.sharding.ClusterShardingSettings.{ StateStoreModeDData, StateStoreModePersistence }
+  require(
+    stateStoreMode == StateStoreModePersistence || stateStoreMode == StateStoreModeDData,
+    s"Unknown 'state-store-mode' [$stateStoreMode], " +
+      s"valid values are '${StateStoreModeDData.name}' or '${StateStoreModePersistence.name}'")
+
+  /** If true, this node should run the shard region, otherwise just a shard proxy should started on this node. */
+  @InternalApi
+  private[akka] def shouldHostShard(cluster: Cluster): Boolean =
+    role.isEmpty || cluster.selfMember.roles(role.get)
+
+  def withRole(role: String): ClusterShardingSettings = copy(role = ClusterShardingSettings.roleOption(role))
+
+  def withRole(role: Option[String]): ClusterShardingSettings = copy(role = role)
+
+  def withRememberEntities(rememberEntities: Boolean): ClusterShardingSettings =
+    copy(rememberEntities = rememberEntities)
+
+  def withJournalPluginId(journalPluginId: String): ClusterShardingSettings =
+    copy(journalPluginId = journalPluginId)
+
+  def withSnapshotPluginId(snapshotPluginId: String): ClusterShardingSettings =
+    copy(snapshotPluginId = snapshotPluginId)
+
+  def withTuningParameters(tuningParameters: ClusterShardingSettings.TuningParameters): ClusterShardingSettings =
+    copy(tuningParameters = tuningParameters)
+
+  def withStateStoreMode(stateStoreMode: ClusterShardingSettings.StateStoreMode): ClusterShardingSettings =
+    copy(stateStoreMode = stateStoreMode)
+
+  /**
+   * The `role` of the `ClusterSingletonManagerSettings` is not used. The `role` of the
+   * coordinator singleton will be the same as the `role` of `ClusterShardingSettings`.
+   */
+  def withCoordinatorSingletonSettings(coordinatorSingletonSettings: ClusterSingletonManagerSettings): ClusterShardingSettings =
+    copy(coordinatorSingletonSettings = coordinatorSingletonSettings)
+
+  private def copy(
+    role:                         Option[String]                           = role,
+    rememberEntities:             Boolean                                  = rememberEntities,
+    journalPluginId:              String                                   = journalPluginId,
+    snapshotPluginId:             String                                   = snapshotPluginId,
+    stateStoreMode:               ClusterShardingSettings.StateStoreMode   = stateStoreMode,
+    tuningParameters:             ClusterShardingSettings.TuningParameters = tuningParameters,
+    coordinatorSingletonSettings: ClusterSingletonManagerSettings          = coordinatorSingletonSettings): ClusterShardingSettings =
+    new ClusterShardingSettings(
+      role,
+      rememberEntities,
+      journalPluginId,
+      snapshotPluginId,
+      stateStoreMode,
+      tuningParameters,
+      coordinatorSingletonSettings)
+}
```

Comment:
The settings are 1:1 but will allow migrating away from using the untyped ones at all in the future.
Code is generated, won't have typos.